### PR TITLE
Learn4haskell chapter4

### PR DESCRIPTION
**Pull request by EspenBerget**
### Solutions for Chapter 4

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +478,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap e) _ = Trap e
+    (<*>) (Reward f) a = fmap f a
```
Comment (chshersh) [inline src/Chapter4.hs:485]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -488,7 +495,17 @@ Implement the 'Applicative' instance for our 'List' type.
...
+append Empty ys = ys
+append (Cons x xs) ys = Cons x (append xs ys) 
+
+instance Applicative List where
+  pure :: a -> List a
+  pure a = Cons a Empty 
 
+  (<*>) :: List (a -> b) -> List a -> List b
+  Empty <*> _ = Empty
+  (Cons f fs) <*> as = append (fmap f as) (fs <*> as)
```
Comment (chshersh) [inline src/Chapter4.hs:508]: Clean and idiomatic implementation :+1:

File: src/Chapter4.hs [hunk 3]
```diff
@@ -629,7 +650,7 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+andM b1 b2 = b1 >>= (\x -> b2 >>= (\y -> pure (x&&y)))
```
Comment (chshersh) [inline src/Chapter4.hs:653]: The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour, so the

```
Just False `andM` _
```

should be `Just False` :slightly_smiling_face:

File: src/Chapter4.hs [hunk 4]
```diff
@@ -672,7 +693,25 @@ Specifically,
...
+
+
+treeToList :: Tree a -> [a] 
+treeToList Leaf = []
+treeToList (Branch a l r) = a : (treeToList l ++ treeToList r)
 
+insert :: Ord a => a -> Tree a -> Tree a 
+insert a Leaf = Branch a Leaf Leaf
+insert a (Branch b l r) | a > b = Branch b (insert a l) r
+insert a (Branch b l r) = Branch b l (insert a r)
```
Comment (chshersh) [inline src/Chapter4.hs:714]: That wasn't necessary, but nice job on implementing the insert function for Binary Search Tree! Usually, it's better implemented using the `compare` function:

```haskell
insert a (Branch b l r) = case compare a b of
    EQ -> ...
    LT -> ...
    GT -> ...
```

so you can handle all 3 cases of comparison explicitly and more visible

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face:

**Comment by EspenBerget** [comment]
Thanks again! Enjoyed the course 👍

**Pull request merged**